lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,253 @@
+        return passed ? 0 : 1
+      end
+
+      def log_possible_errors(maybe_errors)
+        errors = Array(maybe_errors).compact
+        unless errors.empty?
+          @logger.err "Error:"
+          errors.each do |message|
+            @logger.err "    #{message}"
+          end
+          return true
+        end
+      end
+
+
+      # Create crls and keys, then submit them to CA, request for the CA to sign
+      # them, and finally download the signed certificates from the CA.
+      def generate_certs(certnames, settings, digest)
+        name_csr_hash = generate_name_csr_hash(certnames, settings, digest)
+
+        submit_certificate_requests(name_csr_hash, settings)
+        sign_certs(certnames, settings)
+        download_certificates_from_ca(certnames, settings)
+      end
+
+      # Generate a csr and 2 keys for each certname and return a hash in format
+      # of {certname1 => csr1, certname2 => crs2}
+      def generate_name_csr_hash(certnames, settings, digest)
+        certname_csr_hash = {}
+        host = Puppetserver::Ca::Host.new(digest)
+        certnames.each do |certname|
+          new_private_key = host.create_private_key(settings[:keylength])
+          new_public_key = new_private_key.public_key
+          save(new_private_key, certname, settings[:privatekeydir])
+          save(new_public_key, certname, settings[:publickeydir])
+          @logger.inform "Keys for #{certname} saved in #{settings[:privatekeydir]} and #{settings[:publickeydir]}"
+          new_csr = host.create_csr(certname, new_private_key)
+          certname_csr_hash[certname] = new_csr.to_s
+        end
+        certname_csr_hash
+      end
+
+      def save(content, certname, dir)
+        location = File.join(dir, "#{certname}.pem")
+        @logger.warn "#{location} already exists, overwriting" if File.exist?(location)
+        Puppetserver::Utils::FileUtilities.write_file(location, content, 0640)
+      end
+
+
+      # Make an HTTP request to submit certificate requests to CA
+      # @param certname_csr_hash [Hash] hash with certname keys and csr_string values
+      # @param csr [String] string version of a OpenSSL::X509::Request
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] success of all csrs being submitted to CA
+      def submit_certificate_requests(certname_csr_hash, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_request')
+
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certname_csr_hash.each do |certname, csr_string|
+            url.resource_name = certname
+            result = connection.put(csr_string, url)
+
+            log_csr_result(result, certname)
+          end
+        end
+      end
+
+      def log_csr_result(result, certname)
+        case result.code
+        when '200', '204'
```

Comment:
Although this actually comes in as 200 with no content, thought we'd accept both in case they change that to a more standard 204 return in the future.